File: android/app/src/main/java/com/screentime/app/viewmodel/DashboardViewModel.kt
package com.screentime.app.viewmodel

import android.app.Application
import androidx.lifecycle.*
import com.screentime.app.data.model.*
import com.screentime.app.data.repository.RepositoryProvider
import com.screentime.app.data.repository.ScreenTimeRepository
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch

class DashboardViewModel(application: Application) : AndroidViewModel(application) {
    private val repository = RepositoryProvider.getRepository(application)

    val todaySummary: StateFlow<DailyScreenTimeSummary> = repository.getTodaySummary()
        .map { it ?: DailyScreenTimeSummary() }
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), DailyScreenTimeSummary())

    val todayActivities: StateFlow<List<ActivityRecord>> = repository.getTodayActivities()
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())

    val allConfigs: StateFlow<List<AppConfig>> = repository.allAppConfigs
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())

    val allAchievements: StateFlow<List<Achievement>> = repository.allAchievements
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())

    val recentGainsPenalties: StateFlow<List<GainPenaltyEvent>> =
        combine(todayActivities, repository.getTodayNonZeroEntries()) { activities, entries ->
            val events = mutableListOf<GainPenaltyEvent>()

            // Map verified activities with reward to ACTIVITY_REWARD events
            activities
                .filter { it.status == ActivityStatus.VERIFIED && it.rewardEarnedSeconds > 0 }
                .forEach { activity ->
                    events.add(
                        GainPenaltyEvent(
                            id = activity.id + "_act",
                            type = GainPenaltyType.ACTIVITY_REWARD,
                            activityName = activity.displayName,
                            secondsDelta = activity.rewardEarnedSeconds,
                            timestamp = activity.endTime ?: activity.startTime,
                            icon = activity.type.icon
                        )
                    )
                }

            // Map screen time entries with non-zero timeEarnedOrSpentSeconds
            entries.forEach { entry ->
                val isReward = entry.timeEarnedOrSpentSeconds > 0
                events.add(
                    GainPenaltyEvent(
                        id = entry.id + "_entry",
                        type = if (isReward) GainPenaltyType.REWARD_APP else GainPenaltyType.PENALTY_APP,
                        appName = entry.appName,
                        secondsDelta = entry.timeEarnedOrSpentSeconds,
                        timestamp = entry.startTime,
                        icon = if (isReward) "add_circle" else "remove_circle"
                    )
                )
            }

            events.sortedByDescending { it.timestamp }.take(10)
        }
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())

    val motivationalMessage: StateFlow<String> = todaySummary.map { summary ->
        when {
            summary.remainingSeconds <= 0 -> "Screen time limit reached! Complete activities to earn more. 💪"
            summary.remainingSeconds < 1800 -> "Less than 30 minutes left. Consider a break! 🌿"
            summary.totalEarnedSeconds > 0 -> "Great job earning extra time through activities! 🌟"
            else -> "Complete activities to earn bonus screen time! 🎯"
        }
    }.stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), "")

    init {
        viewModelScope.launch {
            repository.initializeDefaultConfigs()
            repository.initializeAchievements()
        }
    }

    fun addEarnedTime(seconds: Long) {
        viewModelScope.launch {
            val current = todaySummary.value
            repository.updateSummary(current.copy(
                totalEarnedSeconds = current.totalEarnedSeconds + seconds
            ))
        }
    }

    fun saveActivityAndEarnTime(activity: ActivityRecord, rewardSeconds: Long) {
        viewModelScope.launch {
            repository.saveActivity(activity)
            val current = todaySummary.value
            repository.updateSummary(current.copy(
                totalEarnedSeconds = current.totalEarnedSeconds + rewardSeconds
            ))
        }
    }

    fun recordUsage(durationSeconds: Long, appConfig: AppConfig?) {
        viewModelScope.launch {
            val current = todaySummary.value
            val penalty = if (appConfig?.configType == AppConfigType.PENALTY) {
                (durationSeconds * kotlin.math.abs(appConfig.minutesPerMinute)).toLong()
            } else 0L
            repository.updateSummary(current.copy(
                totalUsedSeconds = current.totalUsedSeconds + durationSeconds,
                totalPenaltySeconds = current.totalPenaltySeconds + penalty
            ))
        }
    }
}

File: ios/ScreenTime/Sources/ViewModels/DashboardViewModel.swift
import Foundation
import Combine
import SwiftUI

class DashboardViewModel: ObservableObject {
    @Published var todaySummary: DailyScreenTimeSummary = DailyScreenTimeSummary()
    @Published var remainingTimeFormatted: String = "0m"
    @Published var usedTimeFormatted: String = "0m"
    @Published var earnedTimeFormatted: String = "0m"
    @Published var progressValue: Double = 0.0
    @Published var motivationalMessage: String = ""
    @Published var recentActivities: [Activity] = []
    @Published var topPenaltyApps: [AppConfig] = []
    @Published var recentGainsPenalties: [GainPenaltyEvent] = []

    private var cancellables = Set<AnyCancellable>()
    private let dataStore = DataStore.shared
    private let screenTimeService = ScreenTimeService.shared

    init() {
        setupBindings()
        refresh()
    }

    private func setupBindings() {
        dataStore.$todaySummary
            .receive(on: DispatchQueue.main)
            .sink { [weak self] summary in
                self?.todaySummary = summary
                self?.updateFormattedValues()
            }
            .store(in: &cancellables)

        dataStore.$activities
            .receive(on: DispatchQueue.main)
            .sink { [weak self] activities in
                let today = Calendar.current.startOfDay(for: Date())
                let todayActivities = activities
                    .filter { Calendar.current.isDate($0.startTime, inSameDayAs: today) }
                self?.recentActivities = todayActivities
                    .sorted { $0.startTime > $1.startTime }
                    .prefix(3)
                    .map { $0 }
                self?.deriveGainPenaltyEvents(from: todayActivities)
            }
            .store(in: &cancellables)

        dataStore.$appConfigs
            .receive(on: DispatchQueue.main)
            .sink { [weak self] configs in
                self?.topPenaltyApps = configs
                    .filter { $0.configType == .penalty && $0.isEnabled }
                    .sorted { abs($0.minutesPerMinute) > abs($1.minutesPerMinute) }
                    .prefix(3)
                    .map { $0 }
            }
            .store(in: &cancellables)
    }

    private func deriveGainPenaltyEvents(from activities: [Activity]) {
        var events: [GainPenaltyEvent] = []
        for activity in activities where activity.status == .verified && activity.rewardEarned > 0 {
            let event = GainPenaltyEvent(
                id: activity.id.uuidString + "_act",
                type: .activityReward,
                activityName: activity.displayName,
                secondsDelta: Int(activity.rewardEarned),
                timestamp: activity.endTime ?? activity.startTime,
                icon: activity.type.icon
            )
            events.append(event)
        }
        recentGainsPenalties = Array(
            events.sorted { $0.timestamp > $1.timestamp }.prefix(5)
        )
    }

    func refresh() {
        todaySummary = dataStore.todaySummary
        updateFormattedValues()
    }

    private func updateFormattedValues() {
        remainingTimeFormatted = formatTime(todaySummary.remaining)
        usedTimeFormatted = formatTime(todaySummary.totalUsed)
        earnedTimeFormatted = formatTime(todaySummary.totalEarned)
        progressValue = todaySummary.usagePercentage
        motivationalMessage = generateMotivationalMessage()
    }

    private func generateMotivationalMessage() -> String {
        let remaining = todaySummary.remaining
        if remaining <= 0 {
            return "Screen time limit reached! Complete activities to earn more. 💪"
        } else if remaining < 1800 {
            return "Less than 30 minutes left. Consider taking a break! 🌿"
        } else if todaySummary.totalEarned > 0 {
            return "Great job earning extra time through activities! 🌟"
        } else {
            return "Complete activities to earn bonus screen time! 🎯"
        }
    }

    private func formatTime(_ seconds: TimeInterval) -> String {
        let hours = Int(seconds) / 3600
        let minutes = (Int(seconds) % 3600) / 60
        if hours > 0 {
            return "\(hours)h \(minutes)m"
        }
        return "\(minutes)m"
    }

    var progressColor: Color {
        if progressValue < 0.6 {
            return .green
        } else if progressValue < 0.85 {
            return .orange
        } else {
            return .red
        }
    }
}
